plutus-core/untyped-plutus-core/Language/UntypedPlutusCore/Evaluation/Machine/Cek.hs

```diff
@@ -531,9 +535,23 @@ unsafeEvaluateCek runtime = either throw id . extractEvaluationResult . evaluate
 readKnownCek
     :: ( GShow uni, GEq uni, Closed uni, uni `Everywhere` ExMemoryUsage
        , KnownType (Term Name uni fun ()) a
-       , Hashable fun, Ix fun, ExMemoryUsage fun
+       , Hashable fun, Ix fun
        )
     => BuiltinsRuntime fun (CekValue uni fun)
     -> Term Name uni fun ()
     -> Either (CekEvaluationException uni fun) a
 readKnownCek runtime = evaluateCek runtime >=> readKnown
+
+annotateMemory :: (Everywhere uni ExMemoryUsage, Closed uni) =>
+    Term Name uni fun () -> Term Name uni fun ExMemory
+annotateMemory = cata $ \case
```

Comment:
This is fine, but I was just thinking in my head if we could do this nicely with lenses, and I think we could if `toAnnotation` gave us a (type-changing) lens for the annotation.

Something like:
```
... -- Explicit cases for the ones with non-term sub-parts
t -> let subMemory = sumOf (termSubterms . annotationLens) t in set annotationLens subMemory t
```

Reply:
That would actually remove most of them, I'll take a look if I can make it work.

Reply:
I want to know how much the lenses cost! If you make that change, could you run the validation benchmarks before and after to see if there's any noticeable change in the execution time?

Reply:
They're not really possible.